Rails task4 model associations

phantom-profile opened the pull request.

phantom-profile committed:
made associations between categories questions and answers

phantom-profile committed:
made associations between tests and users

1) association as test - test completer
2) association as written test - test author

VidgarVii commented on app/models/category.rb:
```diff
@@ -1,2 +1,5 @@
+# frozen_string_literal: true
+
 class Category < ApplicationRecord
+  has_many :tests
```
если категорию удалить, то в текущем варианте можно в дальнейшем получить ошибку

к примеру при попытке обратиться по ассоциации (`@test.category`)
или при каких-то операциях, которые подразумевают валидацию объекта

после удаления категории в таблице `tests` останется в поле `category_id` ссылка на несуществующий объект
т.е. объекты в таблице `tests` станут невалидными: рельсы по умолчанию проверяют наличие ассоциированного объекта для всех `belongs_to` ассоциаций (при валидации этого объекта) и выбрасывают ошибку валидации, если объекта не существует

если же на уровне бд поле `category_id` обозначено, как внешний ключ, то уже на попытке удаления объекта из таблицы `categories` получим ошибку базы данных, такого примерно вида:
```
ActiveRecord::InvalidForeignKey (PG::ForeignKeyViolation: EROR:  update or delete on table "categories" violates foreign key constraint
```
и рельсы откажутся удалять категорию

поэтому для всех `has_*` ассоциаций следует указывать поведение при удалении - можно посмотреть в документации - опция `dependent: ...`

VidgarVii commented on app/models/question.rb:
```diff
@@ -1,2 +1,7 @@
+# frozen_string_literal: true
+
 class Question < ApplicationRecord
+	has_many :answers
```
1 отступ съехал
2 
Соблюдай очередность.
Принято сначала писать belongs_to далее ниже has_

VidgarVii commented on app/models/test.rb:
```diff
@@ -2,6 +2,12 @@
 # Tests of different categories with questions and answers
 class Test < ApplicationRecord
+  has_many :questions
+  has_and_belongs_to_many :users
```
ассоциация `has_and_belongs_to_many` почти не используется
в 90% случаев или больше используют `has_many through`, потому что ХБТМ довольно негибкий вариант и если вдруг (а это "вдруг" либо почти всегда планируется, либо все равно случается) понадобится большая гибкость и какая-то логика, относящаяся к связи, то потом переводить из одной связи в другую занятие очень затратное
проще изначально использовать ХМФ

(в текущем приложении в дальнейшем понадобится добавлять логику в связь между пользователями и тестами, поэтому точно нужно реализовывать связь через ХМФ)

VidgarVii commented on app/models/test.rb:
```diff
@@ -2,6 +2,12 @@
+  belongs_to :category
+  belongs_to :author, foreign_key: :author_id, class_name: 'User'
+
   def self.by_category(category)
```
с использованием ассоциации `joins` можно сильно сократить
в задании "Ассоциации моделей" самое время это сделать

***
```ruby
  joins(:category)
    .where(categories: { title: category_title })
    .order(title: :desc)
    .pluck(:title)
```

VidgarVii commented on app/models/user.rb:
```diff
@@ -1,9 +1,10 @@
-      .where(level: level, user_tests: { user_id: id })
+    tests.where(level: level)
   end
 end
```
в сидах
теперь, благодаря ассоциациям, можно (и рекомендуется в ООП) передавать объекты для связывания по ключам, вместо аттрибутов: `category: categories[0]`, `author: users[0]` и т.п. (без `_id` , `.id`)

phantom-profile committed:
upgraded models associations

- changed order of belongs_to and has_*
- added :dependent
- removed JOIN-string from Test.by_category
- replaced HBTM to HMT

phantom-profile committed:
removed ids from seeds
